opensearch-project/OpenSearch-Dashboards
# [Data Explorer] Persisted state and syncing, shared context, linking state to component + smaller changes

### Description

This PR might look big but its actually just 3 changes and a few cleanup tasks that simplify the app structure.

Changes:
1. Adds persistent storage using the URL for the following and sets them up to sync 
     1. global state `_g`
     2. app state `_a`
     3. query state `_q`
2.  Adds a shared context in Data Explorer's view service to share context between the panel and canvas view components. Discover uses this to share the data fetching observable.
3. Wired up the doc table, sidenav and the querybar to the application state.

Other minor changes include:
1. renaming `docViewExpand` to `expandedHit` to make it clearer what it contains
2. minor styling fixes for the sidebar
3. Simplified the Doc Table structure by eliminating similar components
4. Added back the DocViewLinks that was deleted in an older commit. (This is still WIP but added to keep the API intact)
5. Adds the initial logic to handle routing changes from legacy discover to data explorer. Currently this only migrates older discover state to data explorer state and is limited to `src/plugins/discover/public/migrate_state.ts` and where ever its referenced.
6. Disabling functional test for now until merge to main since they will anyways fail and will be reactivated before a merge to main.

Note: The commit history has changes that have already been merged but the diff is accurate.

### Issues Resolved

closes #4225 
closes #4227 
closes #4228

## Screenshot

Ignore the layout issues, this will be fixed later

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/20453492/5fdbe524-9ece-43ef-a3a3-c65810eac371


## Testing the changes

To test the state syncing:
1. for global state, modify the time range and refresh to see that the state persists. The data in the doc table should also update
2. for app state, add colums and see the URL update. Again this should persist on refresh and the doc table should change
3. For query state, update the search query and do the same as the above

To test the shared context, updating the columns should update the table and vice versa and only one network request should fire for an update in the netwprk tab to the `opensearch` route

### Check List
